File: api/src/api/schemas/grant.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, model_validator

Tier = Literal["metadata", "reader", "writer"]
AccessMode = Literal["open", "scoped"]


class AccessModeUpdate(BaseModel):
    access_mode: AccessMode


class GrantUpsert(BaseModel):
    user_id: uuid.UUID
    schema_name: str | None = None
    table_name: str | None = None
    tier: Tier

    @model_validator(mode="after")
    def _table_needs_schema(self) -> GrantUpsert:
        if self.table_name is not None and self.schema_name is None:
            raise ValueError("table_name requires schema_name")
        return self


class GrantOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    user_name: str | None = None
    schema_name: str | None
    table_name: str | None
    tier: str
    created_at: datetime


class GrantPrincipalOut(BaseModel):
    """A candidate grant principal — a workspace member, human or service account."""

    user_id: uuid.UUID
    name: str
    email: str
    role: str
    is_service_account: bool


class CatalogGrantsOut(BaseModel):
    access_mode: str
    grants: list[GrantOut]
    principals: list[GrantPrincipalOut]

File: web/src/types/grant.ts
export type GrantTier = "metadata" | "reader" | "writer";
export type AccessMode = "open" | "scoped";

export interface Grant {
  id: string;
  user_id: string;
  user_name: string | null;
  schema_name: string | null;
  table_name: string | null;
  tier: string;
  created_at: string;
}

export interface GrantPrincipal {
  user_id: string;
  name: string;
  email: string;
  role: string;
  is_service_account: boolean;
}

export interface CatalogGrants {
  access_mode: AccessMode;
  grants: Grant[];
  principals: GrantPrincipal[];
}

export interface GrantUpsertInput {
  user_id: string;
  schema_name?: string | null;
  table_name?: string | null;
  tier: GrantTier;
}
